opampserver/pkg/server/handlers.go

```diff
@@ -118,25 +115,27 @@ func (c *ConnectionHandlers) OnAgentToServerMessage(ctx context.Context, request
 }
 
 func (c *ConnectionHandlers) OnConnectionClosed(ctx context.Context, connectionInfo *connection.ConnectionInfo) {
-	c.logger.Info("Connection closed for device", "deviceId", connectionInfo.DeviceId)
-	instrumentationInstanceName := instrumentation_instance.InstrumentationInstanceName(connectionInfo.Pod, int(connectionInfo.Pid))
-	err := c.kubeclient.Delete(ctx, &odigosv1.InstrumentationInstance{
-		TypeMeta: metav1.TypeMeta{
-			APIVersion: "odigos.io/v1alpha1",
-			Kind:       "InstrumentationInstance",
-		},
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      instrumentationInstanceName,
-			Namespace: connectionInfo.Pod.GetNamespace(),
-		},
-	})
-
-	if err != nil && !apierrors.IsNotFound(err) {
-		c.logger.Error(err, "failed to delete instrumentation instance", "instrumentationInstanceName", instrumentationInstanceName)
+	// keep the instrumentation instance CR in unhealthy state so it can be used for troubleshooting
+}
+
+func (c *ConnectionHandlers) OnConnectionNoHeartbeat(ctx context.Context, connectionInfo *connection.ConnectionInfo) error {
+	healthy := false
+	message := fmt.Sprintf("OpAMP server did not receive heartbeat from the agent, last message time: %s", connectionInfo.LastMessageTime.Format("2006-01-02 15:04:05 MST"))
+	// keep the instrumentation instance CR in unhealthy state so it can be used for troubleshooting
+	err := instrumentation_instance.PersistInstrumentationInstanceStatus(ctx, connectionInfo.Pod, connectionInfo.ContainerName, c.kubeclient, connectionInfo.InstrumentedAppName, int(connectionInfo.Pid), c.scheme,
+		instrumentation_instance.WithHealthy(&healthy, common.AgentHealthStatusNoHeartbeat, &message),
+	)
+	if err != nil {
+		return fmt.Errorf("failed to persist instrumentation instance health status on connection timedout: %w", err)
 	}
+
+	return nil
 }
 
 func (c *ConnectionHandlers) PersistInstrumentationDeviceStatus(ctx context.Context, message *protobufs.AgentToServer, connectionInfo *connection.ConnectionInfo) error {
```

Comment:
Eventually this function update the instrumentation instance right?
WDYT about calling it `UpdateInstrumentationInstanceStatus`?

Reply:
good point, renamed the function 👍